### PR TITLE
Improve SDK search error messages

### PR DESCRIPTION
**Pull request by mikolajpp**
On many platforms with non-standard sdk packages current error messages do not give a clue.
This PR improves error messages slightly.

### CODE REVIEW

**Comment by mikolajpp** [comment]
I sign the agreement. 

I don't use google account linked to github, so I am unable to give you any "sign" confirmation except this written post.

**Comment by Hixie** [comment]
@mikolajpp We can't accept a patch unless the GitHub account has explicitly been associated with a Google account that has signed the CLA above. You don't have to have a GitHub account associated with  Google account before starting the process, it handles that for you.

**Comment by mikolajpp** [comment]
Okay, I have signed CLA using a dummy account which is linked to my github name. 
Hope this is enough.

**Comment by Hixie** [comment]
Thanks!

**Comment by Hixie** [comment]
cc @tvolkert or @mravn-google for review.

**Comment by Hixie** [comment]
I'm not sure I really understand the new proposed messages. The old message had a link to a page to download things, which seems like a good thing, but maybe not? It would be better if these messages were all very actionable and clear about what the problem is.

File: packages/flutter_tools/lib/src/android/android_sdk.dart [hunk 1]
```diff
@@ -206,8 +206,11 @@ class AndroidSdk {
     if (!processManager.canRun(adbPath))
       return <String>['Android SDK file not found: $adbPath.'];
 
-    if (sdkVersions.isEmpty || latestVersion == null)
-      return <String>['Android SDK is missing command line tools; download from https://goo.gl/XxQghQ'];
+    if (sdkVersions.isEmpty)
+      return <String>['No Android SDK version found (check SDK for standard layout).'];
```
Comment (mravn-google) [inline packages/flutter_tools/lib/src/android/android_sdk.dart]: What does "standard layout" mean? And what does it mean to check it?
- How does the user verify that the SDK layout is standard?
- What should the user do, if it is not?

**Review by mravn-google — COMMENTED** [review]
This may be an improvement, if an Android SDK is fully installed, but in a "non-standard" way that Flutter tooling does not recognize. I hope we can make the error message it even better (more actionable).

**Comment by mikolajpp** [comment]
Flutter expects specific SDK directory layout which is pretty much supported only by out-of-repo installation via android-studio. All Linux distro I use, other people has been reporting as well, are missing files which flutter is expecting to be present, as well as use different naming structure. 

The way I fixed this error is just by installing sdk via android-studio. 

I have browsed the code detecting the presence of sdk layout, and it is very specific one. Android sdk is listed in requirements, so virtually all failures come from people with sdk installed, but which goes undetected by flutter config code due to different packaging conventions. 

The original error message is confusing, because it is mainly encountered by people with android-sdk already installed (at least from open tickets), and same message is displayed no matter the circumstances (like missing version file). 

Either way, there needs to be better way to indicate this, because it is very frustrating for users (as evidenced by tickets).

File: packages/flutter_tools/lib/src/android/android_sdk.dart (hunk 1 above)
Comment (tvolkert) [reply packages/flutter_tools/lib/src/android/android_sdk.dart]: I think it's clearer to just say where we looked for the versions and what we found.  Something like:

```dart
Iterable<Directory> platforms = <Directory>[]; // android-22, ...
final Directory platformsDir = fs.directory(fs.path.join(directory, 'platforms'));
if (platformsDir.existsSync()) {
  platforms = platformsDir
    .listSync()
    .where((FileSystemEntity entity) => entity is Directory)
    .map<Directory>((FileSystemEntity entity) {
      final Directory dir = entity;
      return dir;
    });
}

StringBuffer msg = new StringBuffer('No valid Android SDK platforms found in ${platformsDir.path}.');
if (platforms.isEmpty) {
  msg.write('  Directory was empty.');
} else {
  msg.write('Candidates were:\n');
  msg.write(platforms
      .map((Directory dir) => '  - ${dir.basename}')
      .join('\n'));
}
return <String>[msg.toString()];
```

Note that the first paragraph above already exists in `_init()` and should be refactored out into `get _platforms` and `get _platformsDir`.

**Comment by Hixie** [comment]
@mikolajpp are you interested in applying some of the suggested edits? No problem if you're not able to, we can take the PR over if you prefer.

**Comment by mikolajpp** [comment]
@Hixie, 

I would definitely prefer for you to take over. Out of time currently :-)

**Comment by Hixie** [comment]
Ok, will do! I've marked the PR so it appears on our radar and @tvolkert will probably end up addressing this in his work. I'll close the PR in the meantime. Thanks!

**Pull request closed without merging**